L2SIVacuum/POUs/Functions/Valves/FB_VFS_2OO3.TcPOU

```diff
@@ -0,0 +1,282 @@
+	q_xClose_C	AT%Q* : BOOL; //3x Valve Close output signals (current req.)
+	q_xOPN_DO	AT%Q* : BOOL; //Valve Open output signal
+	(*3x Fast trigger Inputs from CC Gauge controller.
+	  Make sure all tri CC gauge cards are in the same controller*)
+	i_xTriggerIG1 AT%I* : BOOL; //Fast trigger input from CC gauge 1
+	i_xTriggerIG2 AT%I* : BOOL; //Fast trigger input from CC gauge 2	 
+	i_xTriggerIG3 AT%I* : BOOL; //Fast trigger input from CC gauge 3
+	
+	(*VETO Devices*)
+	i_xVetoValveOpenDO AT%I* : BOOL; (*Link to VFS Interface*)
+	i_xVetoValveClosed AT%I* : BOOL; (*Link to VFS Interface*)
+	
+	(*Interface with external system*) 
+	i_xPress_OK        AT%I* : BOOL; (*Link to VFS Interface, IG.xPRESS_OK, to make sure that the device is connected*)
+	i_xOPN_SW          AT%I* : BOOL; (*Link to VFS Interface, external open signal e.g epics*)
+	i_xCLS_SW          AT%I* : BOOL; (*Link to VFS Interface, external close signal e.g epics*)
+	i_xVAC_FAULT_Reset AT%I* : BOOL; (*Link to VFS Interface Valve Vacuum OK, is set to False when there is Vacuum Fault*)
+	i_xOverrideMode    AT%I* : BOOL; (*Link to VFS Interface, To be linked to global override bit. This Overrides Vacuum logic only, EPS, MPS and PMPS are still enforces*)
+	i_xOverrideOpen    AT%I* : BOOL; (*Link to VFS Interface*)
+	q_xTrigger         AT%Q* : BOOL; (*Link to VFS Interface*)
+	q_xVFS_Open        AT%Q* : BOOL; (*Link to VFS Interface*)
+	q_xVFS_Closed      AT%Q* : BOOL; (*Link to VFS Interface*)
+	q_xVAC_FAULT_OK    AT%Q* : BOOL; (*Link to VFS Interface, Valve Vacuum OK, is set to False when there is Vacuum Fault*)
+	q_xMPS_OK          AT%Q* : BOOL; (*Link to VFS Interface, MPS Fault OK, is set when the Valve is Open and there is no trigger*)
+	q_eVFS_State       AT%Q* : E_VGC; (*Link to VFS Interface*)
+END_VAR]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[(* Fast Shutter/Valve Function Block*)
+(* M.Ghaly Feb 2019*)
+(* J. Govednik Jan 2022 Adding 2OO3 logic*)
+
+(* Fast Shutters do not function as complete isolation valves, but are intended to rapidly limit conductance 
+between vacuum regions.  They should be installed in conjunction with isolation gate valves.  The conductance 
+of the fast shutter in the closed state may be sufficiently low as to create an isolated vacuum region.  
+Thus, at least a pirani gauge, that can measure a pressure down to 1x10-4 Torr, needs to be installed between
+the shutter and the isolation gate valve.
+In the case where a gauge is not installed, the neigboring gate Valves should not be opened unless the shutter 
+is Open. 
+*)
+
+(* On first PLC pass, put valve into vented state, which implies a closed valve *)
+IF xFirstPass THEN
+	fbFF.i_DevName := i_sDevName;
+	xFirstPass := FALSE;
+END_IF
+
+(*set Veto bit*)
+IF (i_xVeto_Enable) THEN
+	xVeto := (i_xVetoValveClosed AND NOT i_xVetoValveOpenDO);
+	ELSE xVeto := FALSE;
+END_IF
+
+(*2oo3 logic*)
+xTrigger := (i_xTriggerIG1 AND i_xTriggerIG2) OR
```

Comment:
Should also check the press ok bit of the triggering gauges. A gauge that is disconnected will never trigger. This is why we use the Press ok bit to make sure the gauge is powered and connected.